File: leetcode/560.和为-k-的子数组.py
#
# @lc app=leetcode.cn id=560 lang=python3
#
# [560] 和为 K 的子数组
#

# @lc code=start
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        mp = defaultdict(int)
        mp[0] = 1
        ans, tmp_sum = 0, 0
        
        for i, v in enumerate(nums):
            tmp_sum += v
            ans += mp[tmp_sum-k]
            mp[tmp_sum] += 1
        return ans
        
# @lc code=end

File: leetcode/560.和为-k-的子数组.cpp
/*
 * @lc app=leetcode.cn id=560 lang=cpp
 *
 * [560] 和为 K 的子数组
 */

// @lc code=start
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        int n = nums.size();
        unordered_map<int, int> cnt;
        cnt[0] = 1;
        int ans = 0;
        int tmp = 0;
        for (int i = 0; i < n; i++) {
            tmp += nums[i];
            int target = tmp-k;
            ans += cnt[target];
            cnt[tmp]++;
        }
        return ans;
    }
};
// @lc code=end

